### PR TITLE
ENH: adding argparsing + unit tests to train.py

### PR DESCRIPTION
**Pull request by sblackburn86**

### CODE REVIEW

File: ecodse_funtime_alpha/data.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def get_dataset(image_dir, labels_csv):
 
     img_label_ds = tf.data.Dataset.zip((img_ds, label_ds))
 
-    return img_label_ds
+    return img_label_ds, len(samples)
```
Comment (hbertrand) [inline ecodse_funtime_alpha/data.py]: No, we should call tf.data.experimental.cardinality

File: ecodse_funtime_alpha/train.py [hunk 2]
```diff
@@ -16,24 +19,49 @@ def train_loop(dataset, model, optimizer):
...
+    if args.seed > 0:
+        tf.random.set_seed(args.seed)
+    dataset, lendataset = data.get_dataset(args.imagepath, args.labelpath)
     # model = models.TestMLP(10, 9)
-    model = models.SimpleCNN(10, 3, 9)
-    optimizer = tf.keras.optimizers.Adam(lr=0.001)
-    fit_loop(dataset, model, optimizer)
+    model = models.SimpleCNN(args.kernels, args.ksize, 9)
+    optimizer = tf.keras.optimizers.Adam(lr=args.lr)
+    fit_loop(dataset, lendataset, model, optimizer, args.nepoch, args.batchsize)
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: Should not  modify get dataset this goes in a separate PR

File: ecodse_funtime_alpha/train.py [hunk 3]
```diff
@@ -16,24 +19,49 @@ def train_loop(dataset, model, optimizer):
...
+                           help='batch size (default=32)')
+    args = argparser.parse_args(args)
+    return args
 
 
 if __name__ == "__main__":
-    image_dir = '/project/def-bengioy/hrb/rainforest/fixed-train-jpg/'
-    labels_csv = '/project/def-bengioy/hrb/rainforest/train_v2.csv'
-    dataset = data.get_dataset(image_dir, labels_csv)
+    tf.enable_eager_execution()
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: You should not enable eager twice. It's already enabled at the top of the file

File: ecodse_funtime_alpha/train.py [hunk 4]
```diff
@@ -16,24 +19,49 @@ def train_loop(dataset, model, optimizer):
...
+    args = get_args(sys.argv[1:])
+    if args.seed > 0:
+        tf.random.set_seed(args.seed)
+    dataset, lendataset = data.get_dataset(args.imagepath, args.labelpath)
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: I don't think that adding a new return to get_dataset in necessary in this PR for argparsr.
Please move this to another PR.

File: ecodse_funtime_alpha/train.py [hunk 5]
```diff
@@ -16,24 +19,49 @@ def train_loop(dataset, model, optimizer):
...
                   metrics=["accuracy"])
-    model.fit(dataset, epochs=20, steps_per_epoch=2)
+    model.fit(dataset, epochs=nepoch, steps_per_epoch=nstep)
+
+
+def get_args(args):
+    argparser = argparse.ArgumentParser()
+    argparser.add_argument('--imagepath', default='../../rainforest/fixed-train-jpg/',
+                           help='path to image directory')
+    argparser.add_argument('--labelpath', default='../../rainforest/train_v3.csv',
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: Even if technically correct, it is common to set the style to everything on one line of one line per hyperparameters and no arbitrary break.

`argparser.add_argument('--imagepath', default='../../rainforest/fixed-train-jpg/', help='path to image directory')`

or

```
    argparser.add_argument('--seed', 
                                             default=-1, 
                                             type=int,
                                             help='Set random seed to this number (if >0)')
```

File: ecodse_funtime_alpha/data.py (hunk 1 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/data.py]: removed from this branch, will implement in a new branch

File: ecodse_funtime_alpha/train.py (hunk 2 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: solved. Change removed from branch, will implement in new branch

File: ecodse_funtime_alpha/train.py (hunk 4 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: moved to another branch.

File: ecodse_funtime_alpha/train.py (hunk 5 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: ok, solved

File: ecodse_funtime_alpha/train.py (hunk 3 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: d'oh! Solved.

File: ecodse_funtime_alpha/train.py [hunk 6]
```diff
@@ -16,24 +19,62 @@ def train_loop(dataset, model, optimizer):
         print(loss)
 
 
-def fit_loop(dataset, model, optimizer):
-    dataset = dataset.shuffle(12)
-    dataset = dataset.repeat(20)
-    dataset = dataset.batch(4)
+def fit_loop(dataset, model, optimizer, nepoch, batchsize):
+    nstep = len(list(dataset)) // batchsize
+    dataset = dataset.shuffle(12)  # 12 = buffer size
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: Advice # 42. When the code can be made clearer don't add a comment for clarification.

dataset = dataset.shuffle(buffer_size=12)

File: ecodse_funtime_alpha/train.py [hunk 7]
```diff
@@ -16,24 +19,62 @@ def train_loop(dataset, model, optimizer):
...
+                           type=int,
+                           help='Number of kernels in the CNN')
+    argparser.add_argument('--ksize',
+                           default=1,
+                           type=int,
+                           help='Size of kernels in CNN')
+    argparser.add_argument('--lr',
+                           default=0.1,
+                           type=float,
+                           help='Learning rate (default=0.1')
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: You should be consistent for putting the default in the help string or not.

If you do you should use a variable to be sure the value is always up to date.
```
default_lr = 0.1
...
...
...
argparser.add_argument('--lr',
                           default=default_lr,
                           type=float,
                           help=f"Learning rate (default={default_lr}")
```

File: ecodse_funtime_alpha/train.py [hunk 8]
```diff
@@ -16,24 +19,62 @@ def train_loop(dataset, model, optimizer):
...
+    args = argparser.parse_args(args)
+    return args
 
 
 if __name__ == "__main__":
-    image_dir = '/project/def-bengioy/hrb/rainforest/fixed-train-jpg/'
-    labels_csv = '/project/def-bengioy/hrb/rainforest/train_v2.csv'
-    dataset = data.get_dataset(image_dir, labels_csv)
+    args = get_args(sys.argv[1:])
+    if args.seed > 0:
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: Why?

File: ecodse_funtime_alpha/train.py [hunk 9]
```diff
@@ -16,24 +19,62 @@ def train_loop(dataset, model, optimizer):
...
+
+def get_args(args):
+    argparser = argparse.ArgumentParser()
+    argparser.add_argument('--imagepath',
+                           default='../../rainforest/fixed-train-jpg/',
+                           help='path to image directory')
+    argparser.add_argument('--labelpath',
+                           default='../../rainforest/train_v3.csv',
+                           help='path to csv file for labels')
+    argparser.add_argument('--seed',
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: You could also add short name.

```
    argparser.add_argument('-s',
                           '--seed',
                           default=-1,
                           type=int,
                           help='Set random seed to this number (if >0)')
```

File: tests/test_train.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        assert args.kernels == 10
+        assert args.ksize == 1
+        assert args.lr == 0.01
+        assert args.nepoch == 2
+        assert args.batchsize == 4
+
+    @pytest.mark.xfail(raises=SystemExit)
+    def test_argparse_lr(self):
+        fakearg = ['--lr=a']
+        _ = get_args(fakearg)
```
Comment (mgermain) [inline tests/test_train.py]: You should assert that the proper exception is thrown.

https://stackoverflow.com/questions/23337471/how-to-properly-assert-that-an-exception-gets-raised-in-pytest

File: ecodse_funtime_alpha/train.py (hunk 6 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: solved.

File: ecodse_funtime_alpha/train.py (hunk 7 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: ok, solved

File: ecodse_funtime_alpha/train.py (hunk 8 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: Removed the if condition.

File: ecodse_funtime_alpha/train.py (hunk 9 above)
Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: done

File: ecodse_funtime_alpha/train.py (hunk 4 above)
Comment (mgermain) [reply ecodse_funtime_alpha/train.py]: You forgot this one.

Comment (mgermain) [reply ecodse_funtime_alpha/train.py]: You forgot to remove `lendataset` here

File: ecodse_funtime_alpha/train.py [hunk 11]
```diff
@@ -16,24 +19,75 @@ def train_loop(dataset, model, optimizer):
         print(loss)
 
 
-def fit_loop(dataset, model, optimizer):
-    dataset = dataset.shuffle(12)
-    dataset = dataset.repeat(20)
-    dataset = dataset.batch(4)
+def fit_loop(dataset, model, optimizer, nepoch, batchsize):
+    nstep = len(list(dataset)) // batchsize
```
Comment (mgermain) [inline ecodse_funtime_alpha/train.py]: Computing the nstep here is unnecessary as it will be done in the next PR. You can leave it to the hard coded `epochs=20, steps_per_epoch=2`

Comment (sblackburn86) [reply ecodse_funtime_alpha/train.py]: fixed
